lib: min_heap: Remove swap() and use byte-wise element swap

sayoojkkarun opened the pull request:
Zephyr does not allow variable-length arrays (VLAs). 
Replace the swap() function which used VLA with a byte-wise swap.

sayoojkkarun committed:
lib: min_heap: Refactor heapify_up/down functions

Remove using `swap()` which was using VLA and replace it
with `byteswp()`.
Reduce the scope of local variables in `heapify_up/down` to
the smallest necessary block.
Replace `while(true)` loop in `heapify_down()` with a bounded `for`
loop.

Signed-off-by: Sayooj K Karun <sayooj@aerlync.com>

kartben reviewed (changes requested):
Check `byteswp()` util :)
https://docs.zephyrproject.org/apidoc/latest/group__sys-util.html#ga8624d1e5411703deac1ab8517f132d7b

kartben reviewed (commented):
it feels like you could probably remove your own swap() utility altogether? 
for heapify_up at least, it's unnecessarily causing calls to min_heap_get_elements for elems you already have.

sayooj-aerlync commented:
> it feels like you could probably remove your own swap() utility altogether? for heapify_up at least, it's unnecessarily causing calls to min_heap_get_elements for elems you already have.

Yes. Possible. I will remove it.

cfriedt commented on lib/min_heap/min_heap.c:
```diff
@@ -22,13 +22,10 @@ LOG_MODULE_REGISTER(min_heap);
  */
 static void swap(struct min_heap *heap, size_t a, size_t b)
```
In case there is another rev of this
```suggestion
 static void swap(const struct min_heap *heap, size_t a, size_t b)
```

sayooj-aerlync replied:
Removed swap()

JarmouniA reviewed (dismissed):
The commit message is no longer accurate, the changes are wider than what the commit's/PR's titles say.

sayooj-aerlync commented:
> > The commit message is no longer accurate, the changes are wider than what the commit's/PR's titles say.
> 
> I will split it to two commits.

Felt a changed commit message is better.

sayooj-aerlync commented:
> Some other observations on min heap right now, and they can be done as separate commits here, but don't need to be
> 
> * `MIN_HEAP_FOREACH()` shouldn't be inside of a `do {} while()` loop and the extra curly braces should be removed. An entry needs to be added to `.clang-format` to get the formatter to format it automatically
> * `MIN_HEAP_DEFINE()` cannot call `min_heap_init()`
> * should `min_heap_get_element()` have an assertion?
> * is `heapify_up()` necessary? it doesn't seem like it compared to the [qsort implementation](https://github.com/zephyrproject-rtos/zephyr/blob/main/lib/libc/minimal/source/stdlib/qsort.c)

I will look into it. But shall I take care of these in another Pull request?

sayooj-aerlync commented:
min-heap

When we insert a new element into a min-heap, we place it at the bottom (the next free leaf).
This can violate the heap property because the new element might be smaller than its parent.
So we use heapify up to move the new element upward until the heap property is restored.

When we remove the minimum element (the root), we replace it with the last element (a leaf).
This new root might be larger than its children, so we fix the heap by heapifying down—pushing
it down to the correct position.

Heap Sort

When building a heap from an unsorted array (bottom-up heap construction),
we start heapifying down from the last parent nodes to the root.
we don’t need heapify up because we are fixing the heap from the top down.

When we extract the min (or max) during heap sort,
we replace the root with the last element and heapify down to restore the heap.
In heap sort, heapify up is usually not needed because we build the heap bottom-up and only adjust by pushing elements down.

sayoojkkarun committed:
include: zephyr: sys: simplify MIN_HEAP_FOREACH macro

Refactor the `MIN_HEAP_FOREACH` macro to use a cleaner
for-loop style removing the need for a third `body` argument.
Update the sample application with the new macro changes.

Signed-off-by: Sayooj K Karun <sayooj@aerlync.com>

sayooj-aerlync commented:
@cfriedt I removed the do while loop and the checkpatch.pl scripts gives me a warning like this:
```
-:49: WARNING:BRACES: braces {} are required around if/while/for/else
#49: FILE: include/zephyr/sys/min_heap.h:234:
+		for (size_t _i = 0; _i < (heap)->size && \
+		    (((node_var) = min_heap_get_element((heap), _i)) || true); \
+		    ++_i)
 
```
This was the reason I went with the `do while()` loop.

sayooj-aerlync commented:
> .checkpatch.conf

checkpatch.pl is throwing the same warning for existing FOREACH macros in the mainline. I just tested couple of them:
SYS_SEM_LOCK:
```
-:222: WARNING:BRACES: braces {} are required around if/while/for/else
#222: FILE: include/zephyr/sys/sem.h:208:
+	for (int __rc SYS_SEM_LOCK_ONEXIT = sys_sem_take((sem), K_FOREVER); ({                     \
+		     __ASSERT(__rc >= 0, "Failed to take sem: %d", __rc);                          \
+		     __rc == 0;                                                                    \
+	     });                                                                                   \
+	     ({                                                                                    \
+		     __rc = sys_sem_give((sem));                                                   \
+		     __ASSERT(__rc == 0, "Failed to give sem: %d", __rc);                          \
+	     }),                                                                                   \
+		      __rc = 1)
+
+/**
+ * @}
```
K_SPINLOCK:
```
-:453: WARNING:BRACES: braces {} are required around if/while/for/else
#453: FILE: include/zephyr/spinlock.h:439:
+	for (k_spinlock_key_t __i K_SPINLOCK_ONEXIT = {}, __key = k_spin_lock(lck); !__i.key;      \
+	     k_spin_unlock((lck), __key), __i.key = 1)
+
+/** @} */
+
```
Do you think the script may need some changes? I have checked for dotfiles as well for additions of macro. only `.clang-format` holds the macros.  

notify @cfriedt

cfriedt commented:
@sayooj-aerlync - is that in upstream CI or locally? If it's upstream I would have thought there would be a lot more noise about it.

sayoojkkarun committed:
include: zephyr: sys: Assert in min_heap_get_element

Assert in `min_heap_get_element()` if heap is NULL

Signed-off-by: Sayooj K Karun <sayooj@aerlync.com>

sayooj-aerlync commented:
@cfriedt
About the comment: `MIN_HEAP_DEFINE() cannot call min_heap_init()`
Shall I remove the macro totally? I think for runtime initialisation, the user can use function call directly.

cfriedt commented on include/zephyr/sys/min_heap.h:
```diff
@@ -225,18 +225,14 @@ static inline void *min_heap_get_element(const struct min_heap *heap,
+ * MIN_HEAP_FOREACH(&heap, node) {
  *	printk("Value: %d\n", node->value);
- * });
+ * };
```
```suggestion
 * }
```

sayooj-aerlync replied:
Done

cfriedt reviewed (commented):
Just a bit, otherwise lgtm

sayoojkkarun committed:
include: zephyr: sys: Refactor MIN_HEAP_DEFINE macro

Remove calling of `min_heap_init()` from MIN_HEAP_DEFINE

Signed-off-by: Sayooj K Karun <sayooj@aerlync.com>

sayooj-aerlync commented:
checkpatch.pl issue is strange. clang-format does not complain about it. I have raised a bug to track it.
https://github.com/zephyrproject-rtos/zephyr/issues/91645

npitre commented:
@keith-zephyr: You assigned this to me but I'm not familiar with this code. My turf is the heap _allocator_.  ;-)

andyross reviewed (approved):
It's actually not true that we never allow alloca/vla usage, there are a few spots where it's extremely useful, though we general provide a fallback for systems that can't allow it (c.f. CONFIG_MISRA_SANE).

In this case the in-place swap is a superior algorithm anyway though.
